File: application/src/main/java/whenyourcar/application/converter/car/CarCommonConverter.java
package whenyourcar.application.converter.car;

import whenyourcar.domain.entity.Car;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.stereotype.Component;
import whenyourcar.domain.query.car.SearchCarsQuery;
import whenyourcar.domain.query.car.SearchDetailCarsQuery;
import whenyourcar.application.dto.car.search.CarCommonResponse;

import java.util.List;
import java.util.stream.Collectors;

@Component
public class CarCommonConverter {
    public Page<CarCommonResponse.SearchResponseDto> toSearchCarsResponse(Page<SearchCarsQuery> searchCarsQueries) {
        List<CarCommonResponse.SearchResponseDto> searchResponses = searchCarsQueries.getContent().stream()
                .map(car -> CarCommonResponse.SearchResponseDto.builder()
                        .carId(car.getCarId())
                        .age(car.getAge())
                        .name(car.getName())
                        .image(car.getImage())
                        .mileage(car.getMileage())
                        .price(car.getPrice())
                        .build())
                .collect(Collectors.toList());
        return new PageImpl<>(searchResponses, searchCarsQueries.getPageable(), searchCarsQueries.getTotalElements());
    }

    public CarCommonResponse.SearchDescriptionResponseDto toDescResponse(Car car) {
        return CarCommonResponse.SearchDescriptionResponseDto.builder()
                .car(car)
                .build();
    }

    public Page<CarCommonResponse.SearchDetailResponseDto> toSearchDetailCarsResponse(Page<SearchDetailCarsQuery> searchDetailCarsQueries) {
        List<CarCommonResponse.SearchDetailResponseDto> searchResponses = searchDetailCarsQueries.getContent().stream()
                .map(car -> CarCommonResponse.SearchDetailResponseDto.builder()
                        .carId(car.getCarId())
                        .age(car.getAge())
                        .name(car.getName())
                        .image(car.getImage())
                        .mileage(car.getMileage())
                        .price(car.getPrice())
                        .build())
                .collect(Collectors.toList());
        return new PageImpl<>(searchResponses, searchDetailCarsQueries.getPageable(), searchDetailCarsQueries.getTotalElements());
    }
}

File: application/src/main/kotlin/howmuchcar/application/converter/car/CarCommonConverter.kt
package howmuchcar.application.converter.car

import howmuchcar.domain.entity.Car
import howmuchcar.domain.query.car.SearchCarsQuery
import howmuchcar.domain.query.car.SearchDetailCarsQuery
import lombok.RequiredArgsConstructor
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.stereotype.Component
import java.util.function.Function
import java.util.stream.Collectors

@Component
@RequiredArgsConstructor
class CarCommonConverter {
    fun toSearchCarsResponse(searchCarsQueries: Page<SearchCarsQuery>): Page<CarSearchResponse> {
        val searchResponses: List<CarSearchResponse> = searchCarsQueries.content
            .map { car: SearchCarsQuery ->
                CarSearchResponse(
                    carId = car.carId,
                    age = car.age,
                    name = car.name,
                    image = car.image,
                    mileage = car.mileage,
                    price = car.price,
                    tag = car.tags
                )
            }
        return PageImpl(searchResponses, searchCarsQueries.pageable, searchCarsQueries.totalElements)
    }

    fun toDescResponse(car: Car): CarSearchDescResponse {
        return CarSearchDescResponse(car = car)
    }

    fun toSearchDetailCarsResponse(searchDetailCarsQueries: Page<SearchDetailCarsQuery>): Page<CarSearchDetailResponse> {
        val searchResponses: List<CarSearchDetailResponse> =
            searchDetailCarsQueries.content
                .map{ car: SearchDetailCarsQuery ->
                    CarSearchDetailResponse(
                        carId = car.carId,
                        age = car.age,
                        name = car.name,
                        image = car.image,
                        mileage = car.mileage,
                        price = car.price,
                        tag = car.tags
                    )
                }
        return PageImpl(
            searchResponses,
            searchDetailCarsQueries.pageable,
            searchDetailCarsQueries.totalElements
        )
    }
}